tesseract/messaging/substrate-evm/src/lib.rs

```diff
@@ -0,0 +1,420 @@
+{
+	async fn query_consensus_state(
+		&self,
+		at: Option<u64>,
+		id: ConsensusStateId,
+	) -> Result<Vec<u8>, Error> {
+		self.evm.query_consensus_state(at, id).await
+	}
+
+	async fn query_latest_height(&self, id: StateMachineId) -> Result<u32, Error> {
+		self.evm.query_latest_height(id).await
+	}
+
+	async fn query_finalized_height(&self) -> Result<u64, Error> {
+		self.evm.query_finalized_height().await
+	}
+
+	async fn query_state_machine_commitment(
+		&self,
+		height: StateMachineHeight,
+	) -> Result<StateCommitment, Error> {
+		self.evm.query_state_machine_commitment(height).await
+	}
+
+	async fn query_state_machine_update_time(
+		&self,
+		height: StateMachineHeight,
+	) -> Result<Duration, Error> {
+		self.evm.query_state_machine_update_time(height).await
+	}
+
+	async fn query_challenge_period(&self, id: StateMachineId) -> Result<Duration, Error> {
+		self.evm.query_challenge_period(id).await
+	}
+
+	async fn query_timestamp(&self) -> Result<Duration, Error> {
+		self.evm.query_timestamp().await
+	}
+
+	async fn query_requests_proof(
+		&self,
+		at: u64,
+		keys: Vec<Query>,
+		_counterparty: StateMachine,
+	) -> Result<Vec<u8>, Error> {
+		let storage_keys: Vec<Vec<u8>> = keys
+			.into_iter()
+			.map(|q| {
+				let slot_hash = tesseract_evm::derive_map_key(
+					q.commitment.0.to_vec(),
+					REQUEST_COMMITMENTS_SLOT,
+				);
+				self.storage_key(slot_hash)
+			})
+			.collect();
+
+		self.fetch_combined_proof(at, storage_keys).await
+	}
+
+	async fn query_responses_proof(
+		&self,
+		at: u64,
+		keys: Vec<Query>,
+		_counterparty: StateMachine,
+	) -> Result<Vec<u8>, Error> {
+		let storage_keys: Vec<Vec<u8>> = keys
+			.into_iter()
+			.map(|q| {
+				let slot_hash = tesseract_evm::derive_map_key(
+					q.commitment.0.to_vec(),
+					RESPONSE_COMMITMENTS_SLOT,
+				);
+				self.storage_key(slot_hash)
+			})
+			.collect();
+
+		self.fetch_combined_proof(at, storage_keys).await
+	}
+
+	async fn query_state_proof(
+		&self,
+		at: u64,
+		keys: StateProofQueryType,
+	) -> Result<Vec<u8>, Error> {
+		let contract_addr: H160 = self.evm.config.ismp_host;
+
+		let storage_keys: Vec<Vec<u8>> = match keys {
+			StateProofQueryType::Ismp(keys) => {
+				if keys.iter().any(|key| key.len() != 32) {
+					return Err(anyhow::anyhow!("All ISMP keys must have a length of 32 bytes",));
+				}
+				keys.into_iter()
+					.map(|key| {
+						let slot = H256::from_slice(&key);
+						self.storage_key(slot)
+					})
+					.collect()
+			},
+			StateProofQueryType::Arbitrary(keys) => {
+				let mut storage_keys = Vec::new();
+				for key in keys.into_iter() {
+					if key.len() != 52 {
+						anyhow::bail!(
+							"All arbitrary keys must have a length of 53 bytes, found {}",
+							key.len()
+						);
+					}
+					let slot = H256::from_slice(&key[20..]);
+					storage_keys.push(self.storage_key(slot));
+				}
+				storage_keys
+			},
+		};
+
+		self.fetch_combined_proof(at, storage_keys).await
```

Comment:
You are meant to fetch a proof for each contract address present in the arbitrary state proof queries, this function should take an array of contract addresses instead of assuming the ismp host address as the contract address for all queries